src/VisualStudio/Core/Def/Implementation/ProjectSystem/FileChangeWatcher.cs

```diff
@@ -149,6 +148,297 @@ public interface IFileWatchingToken
+            /// equivalent to performing the current instance immediately followed by <paramref name="other"/>.
+            /// </summary>
+            /// <param name="other">The next operation to apply.</param>
+            /// <param name="combined">An operation representing the combined application of the current instance and
+            /// <paramref name="other"/>, in that order; otherwise, <see cref="Empty"/> if the current operation cannot
+            /// be combined with <paramref name="other"/>.</param>
+            /// <returns><see langword="true"/> if the current operation can be combined with <paramref name="other"/>;
+            /// otherwise, <see langword="false"/>.</returns>
+            public bool TryCombineWith(in WatcherOperation other, out WatcherOperation combined)
+            {
+                if (other._kind == Kind.None)
+                {
+                    combined = this;
+                    return true;
+                }
+                else if (_kind == Kind.None)
+                {
+                    combined = other;
+                    return true;
+                }
+
+                switch (_kind)
+                {
+                    case Kind.WatchDirectory:
+                    case Kind.WatchFile:
+                        // Watching operations cannot be combined
+                        break;
+
+                    case Kind.UnwatchFile when other._kind == Kind.UnwatchFile:
+                        combined = UnwatchFiles(ImmutableList.Create(_token, other._token));
+                        return true;
+
+                    case Kind.UnwatchFile when other._kind == Kind.UnwatchFiles:
+                        combined = UnwatchFiles(other._tokens.ToImmutableList().Insert(0, _token));
+                        return true;
+
+                    case Kind.UnwatchDirectories when other._kind == Kind.UnwatchDirectories:
+                        var cookies = new List<uint>(_cookies);
+                        cookies.AddRange(other._cookies);
+                        combined = UnwatchDirectories(cookies);
+                        return true;
+
+                    case Kind.UnwatchFiles when other._kind == Kind.UnwatchFile:
+                        combined = UnwatchFiles(_tokens.ToImmutableList().Add(other._token));
+                        return true;
+
+                    case Kind.UnwatchFiles when other._kind == Kind.UnwatchFiles:
+                        combined = UnwatchFiles(_tokens.ToImmutableList().AddRange(other._tokens));
+                        return true;
+
+                    default:
+                        break;
+                }
+
+                combined = default;
+                return false;
+            }
+
+            public async ValueTask ApplyAsync(IVsAsyncFileChangeEx service, CancellationToken cancellationToken)
+            {
+                switch (_kind)
+                {
+                    case Kind.None:
```

Comment:
when does this actually arise in practice?  i dont' mind if this is there for safety, but perhaps we should assert/contract/fatalerror if this really shouldn't happen.  (if it should happen, can you lmk when)?

Reply:
It shouldn't happen today, but in theory it could happen if we ever combined a watch operation with an unwatch operation for the same token.

Reply:
k